File: src/python/modelmesh/cdk/base_secret_store.py
"""Base secret store implementation for the CDK.

Implements the SecretStoreConnector interface with an in-memory
dictionary backend and optional TTL-based caching. Subclasses override
the ``_resolve(name)`` hook to read secrets from files, vaults, or
cloud services.
"""
from __future__ import annotations

import time
from dataclasses import dataclass, field

from modelmesh.interfaces.secret_store import SecretStoreConnector


@dataclass
class BaseSecretStoreConfig:
    """Configuration for a BaseSecretStore instance."""

    secrets: dict[str, str] = field(default_factory=dict)
    cache_enabled: bool = True
    cache_ttl_ms: int = 300_000
    fail_on_missing: bool = True


class BaseSecretStore(SecretStoreConnector):
    """Base implementation of the SecretStoreConnector interface.

    Provides in-memory secret storage with optional TTL-based caching.
    Subclasses override ``_resolve`` to read secrets from external
    backends (files, vaults, cloud services).
    """

    def __init__(self, config: BaseSecretStoreConfig) -> None:
        self._config = config
        self._cache: dict[str, tuple[str, float]] = {}

    def get(self, name: str) -> str:
        """Resolve a secret by name and return its value.

        Checks the cache first (if enabled), then delegates to
        ``_resolve``. Caches the result with a TTL if caching is
        enabled.

        Raises:
            KeyError: If the secret is not found and fail_on_missing
                      is True.
        """
        # Check cache
        if self._config.cache_enabled and name in self._cache:
            value, expires_at = self._cache[name]
            if time.monotonic() < expires_at:
                return value
            del self._cache[name]

        # Resolve from backend
        value = self._resolve(name)
        if value is None:
            if self._config.fail_on_missing:
                raise KeyError(f"Secret not found: {name}")
            return ""

        # Cache the result
        if self._config.cache_enabled:
            expires_at = time.monotonic() + (self._config.cache_ttl_ms / 1000.0)
            self._cache[name] = (value, expires_at)

        return value

    def _resolve(self, name: str) -> str | None:
        """Resolve a secret by name from the configured backend.

        The default implementation looks up the name in the
        ``config.secrets`` dictionary. Override this method to read
        from files, environment variables, vaults, or cloud services.

        Returns:
            The secret value, or None if not found.
        """
        return self._config.secrets.get(name)

    def clear_cache(self) -> None:
        """Clear all cached secret values."""
        self._cache.clear()


__all__ = [
    "BaseSecretStoreConfig",
    "BaseSecretStore",
]

File: src/typescript/src/cdk/base-secret-store.ts
/**
 * Base secret store implementation for the CDK.
 *
 * Implements the SecretStoreConnector interface with an in-memory
 * dictionary backend and optional TTL-based caching. Subclasses override
 * the _resolve(name) hook to read secrets from files, vaults, or
 * cloud services.
 */

import { RuntimeEnvironment } from '../interfaces/runtime';
import { SecretStoreConnector } from '../interfaces/secret-store';

export interface BaseSecretStoreConfig {
  secrets?: Record<string, string>;
  cacheEnabled?: boolean;
  cacheTtlMs?: number;
  failOnMissing?: boolean;
}

export class BaseSecretStore implements SecretStoreConnector {
  static readonly RUNTIME = RuntimeEnvironment.UNIVERSAL;

  protected readonly _config: Required<BaseSecretStoreConfig>;
  private readonly _cache = new Map<string, { value: string; expiresAt: number }>();

  constructor(config?: BaseSecretStoreConfig) {
    this._config = {
      secrets: config?.secrets ?? {},
      cacheEnabled: config?.cacheEnabled ?? true,
      cacheTtlMs: config?.cacheTtlMs ?? 300000,
      failOnMissing: config?.failOnMissing ?? true,
    };
  }

  get(name: string): string {
    // Check cache
    if (this._config.cacheEnabled && this._cache.has(name)) {
      const cached = this._cache.get(name)!;
      if (Date.now() < cached.expiresAt) {
        return cached.value;
      }
      this._cache.delete(name);
    }

    // Resolve from backend
    const value = this._resolve(name);
    if (value == null) {
      if (this._config.failOnMissing) {
        throw new Error('Secret not found: ' + name);
      }
      return '';
    }

    // Cache the result
    if (this._config.cacheEnabled) {
      this._cache.set(name, {
        value,
        expiresAt: Date.now() + this._config.cacheTtlMs,
      });
    }

    return value;
  }

  protected _resolve(name: string): string | null {
    return this._config.secrets[name] ?? null;
  }

  clearCache(): void {
    this._cache.clear();
  }
}
